### PR TITLE
Implement IPSet.Invert and IPSet.Intersect.

### LINKED ISSUE #123: IPSet needs more set operations
While implementing a complicated thing, I would like IPSet.Intersection(a,b) (returns a new IPSet containing only the things in both a and b). And if we're implementing one set operation, the others are probably worth having also (union, symmetric difference).

/cc @bradfitz for thoughts.

### PR DESCRIPTION
**Pull request by danderson**
Fixes #123.

Signed-off-by: David Anderson <dave@natulte.net>

I set the merge-base to #124 so it shows only the new functionality.

### CODE REVIEW

File: ipset.go [hunk 1]
```diff
@@ -99,6 +112,23 @@ func (s *IPSet) RemoveSet(b *IPSet) {
 	}
 }
 
+// Invert updates s to contain the opposite of is current contents.
+func (s *IPSet) Invert() {
+	s.toInOnly()
+	s.out = s.in
+	s.in = []IPRange{
+		IPPrefix{IP: IPv4(0,0,0,0), Bits: 0}.Range(),
```
Comment (mdlayher) [inline ipset.go:120]: Is this gofmt'd? I'd expect spaces between the IPv4 parameters.

**Review by mdlayher — APPROVED** [review]
LGTM!

File: ipset.go (hunk 1 above)
Comment (danderson) [reply ipset.go]: Oops, fixed. My NixOS was missing goimports in the env.

File: ipset.go [hunk 2]
```diff
@@ -22,6 +22,22 @@ type IPSet struct {
...
+	if len(s.out) > 0 {
+		s.in = s.Ranges()
+		s.out = nil
+	}
+}
+
+// Clone returns a copy of s that shares no memory with s.
+func (s *IPSet) Clone() *IPSet {
+	return &IPSet{
+		in:  append([]IPRange(nil), s.in...),
```
Comment (josharian) [inline ipset.go:36]: Could opportunistically do

```go
return &IPSet{in: s.Ranges()}
```

File: ipset.go [hunk 3]
```diff
@@ -99,6 +112,23 @@ func (s *IPSet) RemoveSet(b *IPSet) {
 	}
 }
 
+// Invert updates s to contain the opposite of is current contents.
```
Comment (josharian) [inline ipset.go:115]: Consider calling this `Complement`, since that's the standard set theory name for it.

s/is/its/

File: ipset.go [hunk 4]
```diff
@@ -99,6 +112,23 @@ func (s *IPSet) RemoveSet(b *IPSet) {
...
+func (s *IPSet) Invert() {
+	s.toInOnly()
+	s.out = s.in
+	s.in = []IPRange{
+		IPPrefix{IP: IPv4(0, 0, 0, 0), Bits: 0}.Range(),
+		IPPrefix{IP: IPv6Unspecified(), Bits: 0}.Range(),
```
Comment (josharian) [inline ipset.go:121]: Should we iterate through the `s.in`, discover all IPv6 zones contained therein, and add those here? Should we have IPv6 at all if `s.in` only contains IPv4?

Either way, we should document what universe we're taking the complement in.

File: ipset.go (hunk 3 above)
Comment (danderson) [reply ipset.go:115]: Doh, I knew I was using the wrong word. Thanks!

File: ipset.go (hunk 4 above)
Comment (danderson) [reply ipset.go:121]: Hmmm. IPSet doesn't really deal with zones at all right now (e.g. compaction will smush together IPs across zones), so I think I'm going to turn this into a bug to figure out what to do, and we can fix all of IPSet at once. Until then, leaving as-is.

**Pull request merged**